Adds libvirt staging environment for Focal

conorsch opened the pull request:
## Status

Ready for review.

Towards #5468.

## Description of Changes

Early stages of Focal support for staging VMs. It's partial progress—areas not currently working are:

* drivers other than libvirt (no virtualbox, no qubes support for staging env)
* kernel tasks must be skipped
* testinfra tests are not passing (mostly kernel tasks & version strings, e.g. python paths)


## Testing

First, prepare the boxes for libvirt, same as required for Xenial:


```
vagrant box add bento/ubuntu-20.04
vagrant box mutate bento/ubuntu-20.04 libvirt
molecule converge -s libvirt-staging-focal -- --skip-tags grsecurity
```

Then browse to the Source Interface and interact with the app. Report any problems.

## Deployment
Yes, provisioning logic has changed a bit. An attempt was made to alter only Focal-related logic where possible, but in two places in particular Xenial-related logic was changed:

* the `aptitude` tasks for deciding whether to mark `tor` as held now uses failed_when=false, for Focal compatibility
* the base apt dependencies in the "common" role now use dist-specific vars, to handle Xenial & Focal differently. 

## Checklist

### If you made changes to the server application code:

- [ ] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made changes to `securedrop-admin`:

- [ ] Linting and tests (`make -C admin test`) pass in the admin development container

### If you made changes to the system configuration:

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [ ] I have written a test plan and validated it for this PR

### If you made changes to documentation:

- [ ] Doc linting (`make docs-lint`) passed locally

### If you added or updated a code dependency:

Choose one of the following:

- [ ] I have performed a diff review and pasted the contents to [the packaging wiki](https://github.com/freedomofpress/securedrop-debian-packaging/wiki)
- [ ] I would like someone else to do the diff review

Commit:
Begins libvirt-staging-focal scenario

Starting with a naive copy/paste from the libvirt-staging-xenial
scenario. Will tweak settings as required to get it to run.

Commit:
Focal scenario: updates references xenial -> focal

The bare minimum required to get the copy/pasted Xenial libvirt scenario
working under Focal. Commiting separately to make the changes a bit
easier to inspect.

Commit:
Uses platform-specific vars in common role

In order to ease the transition to Focal, let's use different vars for
packages depending on platform. When running the Xenial logic against
Focal, the kernel removal tasks failed since several of the package
names were not found.

Also adding a "grsecurity" tag since the grsec story for Focal still
needs work, so it's easy to skip now.

kushaldas commented:
This is now blocked on https://github.com/freedomofpress/securedrop/pull/5443

kushaldas commented:
Tested this draft PR using the builder image in #5484 + the osseg dependency fix from this PR, and #5489 PR for symlink. I can build a proper staging on Focal (- grsec)

rmol committed:
Update tor upgrade check for newer aptitude

Between the version of aptitude on Xenial and that on Focal, empty
searches started returning 1, causing `Check whether tor will be
upgraded` to fail. Let's ignore the exit code, since what we care
about is whether "tor" is in the stdout.

We could use `apt list --upgradable`, but that command warns about
lack of a stable CLI, and additionally we'd have to munge the output
further to inspect it.

Commit:
Adds "staging-focal" target

Configures a libvirt VM staging environment based on Focal.
Far from complete, but it's a start.

kushaldas reviewed (approved):
## Testing

- [x] vagrant box add bento/ubuntu-20.04
- [x] vagrant box mutate bento/ubuntu-20.04 libvirt
- [x] molecule converge -s libvirt-staging-focal -- --skip-tags grsecurity
- [x] Can see the source interface 

This is ready to merged in the current state.
